### DISCUSSION TITLE
Enable URL encoding for the name part of --data-urlencode

### DISCUSSION THREAD

**Discussion by ghost** [curl/curl#6667] [Ideas]
We noticed that the name part of a parameter for `--data-urlencode` always needs to be URL-encoded beforehand.
Therefore the following program call would fail:

`curl --data-urlencode 'form-data[0]=example' https://www.example.com`

Also we stumbled upon the fact that the `--form` parameter can actually handle complicated cases according to the man page if values contain characters with special meanings:

> If filename/path contains ',' or ';', it must be quoted by double-quotes like:
  curl -F "file=@\"localfile\";filename=\"nameinpost\"" example.com
  or
  curl -F 'file=@"localfile";filename="nameinpost"' example.com

So, to allow URL encoding of names something like this can be thought of:

`curl --data-urlencode '"form-data[0]"="example"' https://www.example.com`

An advantage would be a clearer reading of what is happening. This also removes the need of preprocessing the names for `--data` parameters.

What to do you think?

**Comment by bagder** [comment]
If we can come up with a syntax that has no or at least minimal risk of breaking existing scripts I wouldn't mind a way to specify it and encode the name too. I don't think using double-quotes like that matches that criteria.

There's also a problem if the name contains a `=` of course as then how do you know when the name ends and the contents begin? IIRC that's the main reason I never implemented this thing in the first place.

**Comment by ghost** [comment]
Thanks for your quick response. Actually it was thought of that double-quotes might also help better dividing the name from the value part. Therefore, something like the following may be possible:

`curl --data-urlencode '"form-data=mark[0]"="example"' https://www.example.com`

However, it is understandable that there is a risk that existing scripts would break. Also, using more and more quotes do not help concerning readability. Maybe this discussion is useful to anyone who comes across the same thought.
